### PR TITLE
metrics: Deprecate non-conventional prometheus metrics

### LINKED ISSUE #11743: add a lint for cilium metrics
This can be achieved by starting cilium with `--prometheus-serve-addr=:9090`
```
$ curl  -s http://localhost:9090/metrics  | ./promtool check metrics 
cilium_bpf_map_ops_total label names should be written in 'snake_case' not 'camelCase'
cilium_bpf_map_ops_total label names should be written in 'snake_case' not 'camelCase'
cilium_bpf_map_ops_total label names should be written in 'snake_case' not 'camelCase'
cilium_bpf_map_ops_total label names should be written in 'snake_case' not 'camelCase'
cilium_bpf_map_ops_total label names should be written in 'snake_case' not 'camelCase'
cilium_bpf_map_ops_total label names should be written in 'snake_case' not 'camelCase'
cilium_bpf_map_ops_total label names should be written in 'snake_case' not 'camelCase'
cilium_bpf_map_ops_total label names should be written in 'snake_case' not 'camelCase'
cilium_bpf_map_ops_total label names should be written in 'snake_case' not 'camelCase'
cilium_bpf_map_ops_total label names should be written in 'snake_case' not 'camelCase'
cilium_endpoint_count non-histogram and non-summary metrics should not have "_count" suffix
cilium_endpoint_regenerations counter metrics should have "_total" suffix
cilium_identity_count non-histogram and non-summary metrics should not have "_count" suffix
cilium_k8s_client_api_calls_counter counter metrics should have "_total" suffix
cilium_k8s_client_api_calls_counter metric name should not include type 'counter'
cilium_nodes_all_events_received_total label names should be written in 'snake_case' not 'camelCase'
cilium_nodes_all_events_received_total label names should be written in 'snake_case' not 'camelCase'
cilium_policy_count non-histogram and non-summary metrics should not have "_count" suffix
cilium_policy_import_errors counter metrics should have "_total" suffix
```

promtool is can be build from https://github.com/prometheus/prometheus/tree/master/cmd/promtool

We can also add the promtool inside the vendor directory similar to what with do with the k8s code generators. For example:
https://github.com/cilium/cilium/blob/3a1bde594b61b1e3425a216b8ae4f7c84b0b8136/tools/go-to-protobuf#L1-L5
https://github.com/cilium/cilium/blob/2b460d7b0a1607e5618492bce38a158ba8f3fb1c/tools/tools.go#L19-L29
https://github.com/cilium/cilium/blob/2b460d7b0a1607e5618492bce38a158ba8f3fb1c/Makefile#L114-L115

### PR DESCRIPTION
**Pull request by sayboras**
metrics: Add new prometheus metrics and related label

This PR is to add new metrics and labels as per the check done
by official tool promtool.

Summary of the changes can be found as below:
- Deprecate eventType, mapName labels in favour of event_type and map_name
- Add _total suffix for counter metrics (cilium_endpoint_regenerations,
cilium_k8s_client_api_calls_counter, cilium_policy_import_errors)
- Add _count suffix for non-histogram and non-summary metrics
(cilium_identity_count, cilium_policy_count)

Fixes: #11743
Signed-off-by: Tam Mach <sayboras@yahoo.com>

```release-note
metrics: Deprecate non-conventional prometheus metrics
```

### CODE REVIEW

**Comment by aditighag** [comment]
Looks like we need to add a note regarding upgrade impact?

**Review by aditighag — APPROVED** [review]

**Comment by sayboras** [comment]
> Looks like we need to add a note regarding upgrade impact?

Thanks, I add some details in upgrade.rst, still not 100% sure if it's enough. Keen to hear your inputs as well.

**Comment by sayboras** [comment]
test-me-please

**Comment by sayboras** [comment]
retest-gke

File: Documentation/configuration/metrics.rst
Comment (fristonio) [inline Documentation/configuration/metrics.rst]: I believe this should also be changed to `endpoint`, similar to how `policy_count` was changed to `policy`?

File: pkg/endpointmanager/manager.go
Comment (fristonio) [inline pkg/endpointmanager/manager.go:170]: One question, I don't see this metric `endpoint_count`/`endpoint` in the default metrics list https://github.com/cilium/cilium/pull/12826/files#diff-7cbe8e056d62a2de30c7066e359bd9c9R461 or grafana and prometheus example config. Is there a reason for this? cc @aanm

**Review by fristonio — CHANGES_REQUESTED** [review]

File: Documentation/configuration/metrics.rst
Comment (sayboras) [reply Documentation/configuration/metrics.rst]: good pick, I missed for both :)

**Comment by sayboras** [comment]
CI passed for all, except GKE due to `‘jenkins-node-gke-fixed-1’ is offline`.

Force push to update metrics.rst as per review comment.

**Review by aanm — CHANGES_REQUESTED** [review]
@sayboras thank you for these changes. I gave a little more thought and it's probably better if we keep the old prometheus metrics for the 1.9 and deleted them in 1.10.

a) Can we have 2 metrics that report the same values?
b) Perhaps we could also have an option to not expose those deprecated metrics?


Also, we would need to add upgrade notes as we are deprecating those metrics and they will be removed in 1.10.

**Comment by sayboras** [comment]
> it's probably better if we keep the old prometheus metrics for the 1.9 and deleted them in 1.10.
> 
> a) Can we have 2 metrics that report the same values?
> b) Perhaps we could also have an option to not expose those deprecated metrics?
> 
> Also, we would need to add upgrade notes as we are deprecating those metrics and they will be removed in 1.10.

Oh I see, it definitely helps avoiding breaking changes. Let me mark this PR as draft, will make changes later then.

**Comment by sayboras** [comment]
test-me-please

File: Documentation/configuration/metrics.rst
Comment (aanm) [inline Documentation/configuration/metrics.rst]: "(deprecated, use ``endpoint_regenerations_total``)"

File: Documentation/configuration/metrics.rst
Comment (aanm) [inline Documentation/configuration/metrics.rst]: "(deprecated, use ``policy``)"

File: Documentation/configuration/metrics.rst
Comment (aanm) [inline Documentation/configuration/metrics.rst]: similar to the previous point it out.

File: Documentation/configuration/metrics.rst
Comment (aanm) [inline Documentation/configuration/metrics.rst]: similar to the previous point it out.

File: Documentation/install/upgrade.rst
Comment (aanm) [inline Documentation/install/upgrade.rst]: please give an example, it was not clear at a first sight what was the difference between `eventType` and `event_type`

File: pkg/bpf/map_linux.go
Comment (aanm) [inline pkg/bpf/map_linux.go]: I believe you meant `remove in 1.10`

File: pkg/bpf/map_linux.go
Comment (aanm) [inline pkg/bpf/map_linux.go]: I believe you meant `remove in 1.10`

File: pkg/endpointmanager/manager.go
Comment (aanm) [inline pkg/endpointmanager/manager.go]: I believe you meant `remove in 1.10`

File: pkg/endpointmanager/manager.go
Comment (aanm) [reply pkg/endpointmanager/manager.go:170]: no, it's likely forgotten

File: pkg/k8s/watchers/watcher.go
Comment (aanm) [inline pkg/k8s/watchers/watcher.go]: I believe you meant `remove in 1.10`

File: pkg/metrics/metrics.go
Comment (aanm) [inline pkg/metrics/metrics.go]: I believe you meant `remove in 1.10`

File: pkg/metrics/metrics.go
Comment (aanm) [inline pkg/metrics/metrics.go]: I believe you meant `remove in 1.10`

File: pkg/metrics/metrics.go
Comment (aanm) [inline pkg/metrics/metrics.go]: I believe you meant `remove in 1.10`

File: pkg/metrics/metrics.go
Comment (aanm) [inline pkg/metrics/metrics.go]: I believe you meant `remove in 1.10`

File: pkg/metrics/metrics.go
Comment (aanm) [inline pkg/metrics/metrics.go]: I believe you meant `remove in 1.10`

File: pkg/metrics/metrics.go
Comment (aanm) [inline pkg/metrics/metrics.go]: I believe you meant `remove in 1.10`

File: pkg/metrics/metrics.go
Comment (aanm) [inline pkg/metrics/metrics.go]: I believe you meant `remove in 1.10`

File: pkg/metrics/metrics.go
Comment (aanm) [inline pkg/metrics/metrics.go]: please add `use ... ` instead

File: pkg/metrics/metrics.go
Comment (aanm) [inline pkg/metrics/metrics.go:677]: This was probably supposed to be part of the previous commit?

**Review by aanm — CHANGES_REQUESTED** [review]
commit looks good, only minor nits.

There are a couple places I wrote 'I believe you meant `remove in 1.10`" but I didn't write on all of the,.

Also, the help messages of each metrics being deprecated should probably contain the "use foo_bar instead"

File: Documentation/install/upgrade.rst
Comment (tklauser) [inline Documentation/install/upgrade.rst]: Please mention the version these will be removed, i.e. 1.10. Maybe something like

```
* ``foobar`` is deprecated and will be removed in Cilium 1.10. Please use ``bazbar`` instead.
```

File: pkg/metrics/metrics.go
Comment (sayboras) [reply pkg/metrics/metrics.go]: my bad :sweat_smile:

File: Documentation/install/upgrade.rst
Comment (sayboras) [reply Documentation/install/upgrade.rst]: thanks :100:

**Comment by sayboras** [comment]
test-me-please

File: .github/workflows/smoke-test.yaml
Comment (fristonio) [inline .github/workflows/smoke-test.yaml]: Does it make sense to only dump all the metrics if we have a failure in `promtool check metrics`?

**Review by fristonio — APPROVED** [review]
This looks great. Just a minor question, does not need a second review from me.
:100:

**Review by tklauser — APPROVED** [review]

**Review by aanm — APPROVED** [review]

**Review by vadorovsky — APPROVED** [review]

File: .github/workflows/smoke-test.yaml
Comment (christarazi) [inline .github/workflows/smoke-test.yaml]: We could remove the usage of cat, and reference the file directly in `egrep`.

```suggestion
          egrep -v "${{ env.DEPRECATED_METRICS }}" metrics.prom | $HOME/go/bin/promtool check metrics
```

File: .github/workflows/smoke-test.yaml
Comment (christarazi) [reply .github/workflows/smoke-test.yaml]: +1, to me it makes sense to dump the metrics if `promtool` has caught something.

**Review by christarazi — APPROVED** [review]
LGTM, minor item to address, no need for a second review.

File: .github/workflows/smoke-test.yaml
Comment (sayboras) [reply .github/workflows/smoke-test.yaml]: :100:

**Comment by sayboras** [comment]
Incorporate review comments from Deepak and Chris. Full CI is not required.

<details>
<summary>git diff</details>

```
diff --git a/.github/workflows/smoke-test.yaml b/.github/workflows/smoke-test.yaml
index 1fc35a5be..582a63b1f 100644
--- a/.github/workflows/smoke-test.yaml
+++ b/.github/workflows/smoke-test.yaml
@@ -161,8 +161,7 @@ jobs:
           kubectl -n kube-system exec $cilium_pod -- sh -c "apt update && apt install curl -y"
           kubectl -n kube-system exec $cilium_pod -- curl http://localhost:9090/metrics > metrics.prom
           go get -u github.com/prometheus/prometheus/cmd/promtool
-          cat metrics.prom
-          cat metrics.prom | egrep -v "${{ env.DEPRECATED_METRICS }}" | $HOME/go/bin/promtool check metrics
+          egrep -v "${{ env.DEPRECATED_METRICS }}" metrics.prom | $HOME/go/bin/promtool check metrics
 
       - name: Dump cilium related logs and events
         if: ${{ failure() }}
```

</details>

**Comment by sayboras** [comment]
Consider that all of code owner approved. There was one small change in helm just cleaning up (e.g. remove null creationTimestamp), so helm team approval is optional. CI also passed.

I am marking this one ready to merge.

**Pull request merged**
